File: backend/app/api/routes/charts.py
"""API routes for Auto-Chart Generation."""
from __future__ import annotations

from typing import Any, Dict, List, Optional
from pydantic import BaseModel, Field
from fastapi import APIRouter, Depends, Request

from backend.app.core.security import require_api_key
from backend.app.domain.charts.service import AutoChartService

router = APIRouter(dependencies=[Depends(require_api_key)])


class ChartAnalyzeRequest(BaseModel):
    data: List[Dict[str, Any]] = Field(..., min_items=1, max_items=100)
    column_descriptions: Optional[Dict[str, str]] = None
    max_suggestions: int = Field(default=3, ge=1, le=10)


class ChartGenerateRequest(BaseModel):
    data: List[Dict[str, Any]] = Field(..., min_items=1, max_items=1000)
    chart_type: str
    x_field: str
    y_fields: List[str]
    title: Optional[str] = None


def get_service() -> AutoChartService:
    return AutoChartService()


@router.post("/analyze")
async def analyze_for_charts(
    payload: ChartAnalyzeRequest,
    request: Request,
    svc: AutoChartService = Depends(get_service),
):
    """Analyze data and suggest appropriate chart visualizations."""
    correlation_id = getattr(request.state, "correlation_id", None)
    suggestions = svc.analyze_data_for_charts(
        data=payload.data,
        column_descriptions=payload.column_descriptions,
        max_suggestions=payload.max_suggestions,
        correlation_id=correlation_id,
    )
    return {"status": "ok", "suggestions": suggestions, "correlation_id": correlation_id}


@router.post("/generate")
async def generate_chart_config(
    payload: ChartGenerateRequest,
    request: Request,
    svc: AutoChartService = Depends(get_service),
):
    """Generate a chart configuration."""
    correlation_id = getattr(request.state, "correlation_id", None)
    config = svc.generate_chart_config(
        data=payload.data,
        chart_type=payload.chart_type,
        x_field=payload.x_field,
        y_fields=payload.y_fields,
        title=payload.title,
    )
    return {"status": "ok", "chart": config, "correlation_id": correlation_id}

File: frontend/src/api/charts.js
/**
 * Auto-Chart Generation API Client
 */
import apiClient from './client';

/**
 * Analyze data for chart recommendations
 * @param {Object} params
 * @param {Array} params.data - Data rows to analyze
 * @param {Object} [params.columnDescriptions] - Optional column descriptions
 * @param {number} [params.maxSuggestions=3] - Max number of suggestions
 */
export async function analyzeData({ data, columnDescriptions, maxSuggestions = 3 }) {
  const response = await apiClient.post('/charts/analyze', {
    data,
    column_descriptions: columnDescriptions || undefined,
    max_suggestions: maxSuggestions,
  });
  return response.data;
}

/**
 * Generate chart configuration
 * @param {Object} params
 * @param {Array} params.data - Data rows
 * @param {string} params.chartType - Chart type (bar, line, pie, etc.)
 * @param {string} params.xField - X-axis field name
 * @param {string[]} params.yFields - Y-axis field names (array)
 * @param {string} [params.title] - Chart title
 */
export async function generateChart({ data, chartType, xField, yFields, title }) {
  const response = await apiClient.post('/charts/generate', {
    data,
    chart_type: chartType,
    x_field: xField,
    y_fields: Array.isArray(yFields) ? yFields : [yFields],
    title,
  });
  return response.data;
}

export default {
  analyzeData,
  generateChart,
};
